### ISSUE TITLE
Race condition / use-after-free in ResourceLoader::load_threaded_get

### ISSUE THREAD

**Issue by alula** [godotengine/godot#63461]
### Godot version

4.0.dev (72b5a4335e914997ef46928034e8371724b29280)

### System information

Windows 11, Vulkan, RTX 3060 

### Issue description

```
The thread 0x3f7c has exited with code 0 (0x0).
    #0 0x7ff7b2fc0712 in ResourceLoader::load_threaded_get C:\Users\alula\workspace\godot\core\io\resource_loader.cpp:491
    #1 0x7ff7abb192aa in ResourceLoaderText::_parse_ext_resource C:\Users\alula\workspace\godot\scene\resources\resource_format_text.cpp:162
    #2 0x7ff7abb7494f in ResourceLoaderText::_parse_ext_resources C:\Users\alula\workspace\godot\scene\resources\resource_format_text.h:82
    #3 0x7ff7b52940eb in VariantParser::parse_value C:\Users\alula\workspace\godot\core\variant\variant_parser.cpp:900
    #4 0x7ff7b52a0336 in VariantParser::parse_tag_assign_with_value_func_eof C:\Users\alula\workspace\godot\core\variant\variant_parser.cpp:1442
    #5 0x7ff7b52a1178 in VariantParser::parse_tag_assign_eof C:\Users\alula\workspace\godot\core\variant\variant_parser.cpp:1452
    #6 0x7ff7abb1d0ba in ResourceLoaderText::_parse_node_tag C:\Users\alula\workspace\godot\scene\resources\resource_format_text.cpp:283
    #7 0x7ff7abb2db25 in ResourceLoaderText::load C:\Users\alula\workspace\godot\scene\resources\resource_format_text.cpp:788
    #8 0x7ff7abb59971 in ResourceFormatLoaderText::load C:\Users\alula\workspace\godot\scene\resources\resource_format_text.cpp:1697
    #9 0x7ff7b2fb9478 in ResourceLoader::_load C:\Users\alula\workspace\godot\core\io\resource_loader.cpp:223
    #10 0x7ff7b2fba5a3 in ResourceLoader::_thread_load_function C:\Users\alula\workspace\godot\core\io\resource_loader.cpp:250
    #11 0x7ff7b2920fa7 in Thread::callback C:\Users\alula\workspace\godot\core\os\thread.cpp:75
    #12 0x7ff7b2924e18 in std::invoke<void (*)(Thread *, const Thread::Settings &, void (*)(void *), void *),Thread *,Thread::Settings,void (*)(void *),void *> C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31627\include\type_traits:1558
    #13 0x7ff7b292305d in std::thread::_Invoke<std::tuple<void (*)(Thread *, const Thread::Settings &, void (*)(void *), void *),Thread *,Thread::Settings,void (*)(void *),void *>,0,1,2,3,4> C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31627\include\thread:55
    #14 0x7ff7b6d7f909 in thread_start<unsigned int (__cdecl*)(void *),1> minkernel\crts\ucrt\src\appcrt\startup\thread.cpp:97
    #15 0x7ff7b6ce1719 in __asan::AsanThread::ThreadStart D:\a\_work\1\s\src\vctools\asan\llvm\compiler-rt\lib\asan\asan_thread.cpp:262
    #16 0x7ffa3c41244c in BaseThreadInitThunk+0x1c (C:\WINDOWS\System32\KERNEL32.DLL+0x18001244c)
    #17 0x7ffa3c70df77 in RtlUserThreadStart+0x27 (C:\WINDOWS\SYSTEM32\ntdll.dll+0x18005df77)

0x11e2654defc8 is located 168 bytes inside of 176-byte region [0x11e2654def20,0x11e2654defd0)
freed by thread T80 here:
    #0 0x7ff7b6cf1934 in free D:\a\_work\1\s\src\vctools\asan\llvm\compiler-rt\lib\asan\asan_malloc_win.cpp:112
    #1 0x7ff7b272a990 in Memory::free_static C:\Users\alula\workspace\godot\core\os\memory.cpp:168
    #2 0x7ff7b2fdffe7 in memdelete<Semaphore> C:\Users\alula\workspace\godot\core\os\memory.h:111
    #3 0x7ff7b2fbab7c in ResourceLoader::_thread_load_function C:\Users\alula\workspace\godot\core\io\resource_loader.cpp:274
    #4 0x7ff7b2920fa7 in Thread::callback C:\Users\alula\workspace\godot\core\os\thread.cpp:75
    #5 0x7ff7b2924e18 in std::invoke<void (*)(Thread *, const Thread::Settings &, void (*)(void *), void *),Thread *,Thread::Settings,void (*)(void *),void *> C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31627\include\type_traits:1558
    #6 0x7ff7b292305d in std::thread::_Invoke<std::tuple<void (*)(Thread *, const Thread::Settings &, void (*)(void *), void *),Thread *,Thread::Settings,void (*)(void *),void *>,0,1,2,3,4> C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31627\include\thread:55
    #7 0x7ff7b6d7f909 in thread_start<unsigned int (__cdecl*)(void *),1> minkernel\crts\ucrt\src\appcrt\startup\thread.cpp:97
    #8 0x7ff7b6ce1719 in __asan::AsanThread::ThreadStart D:\a\_work\1\s\src\vctools\asan\llvm\compiler-rt\lib\asan\asan_thread.cpp:262
    #9 0x7ffa3c41244c in BaseThreadInitThunk+0x1c (C:\WINDOWS\System32\KERNEL32.DLL+0x18001244c)
    #10 0x7ffa3c70df77 in RtlUserThreadStart+0x27 (C:\WINDOWS\SYSTEM32\ntdll.dll+0x18005df77)

previously allocated by thread T75 here:
    #0 0x7ff7b6cf1a84 in malloc D:\a\_work\1\s\src\vctools\asan\llvm\compiler-rt\lib\asan\asan_malloc_win.cpp:121
    #1 0x7ff7b2729b4e in Memory::alloc_static C:\Users\alula\workspace\godot\core\os\memory.cpp:75
    #2 0x7ff7b2729a29 in operator new C:\Users\alula\workspace\godot\core\os\memory.cpp:40
    #3 0x7ff7b2fbc592 in ResourceLoader::load_threaded_request C:\Users\alula\workspace\godot\core\io\resource_loader.cpp:386
    #4 0x7ff7abb26d2a in ResourceLoaderText::load C:\Users\alula\workspace\godot\scene\resources\resource_format_text.cpp:468
    #5 0x7ff7abb59971 in ResourceFormatLoaderText::load C:\Users\alula\workspace\godot\scene\resources\resource_format_text.cpp:1697
    #6 0x7ff7b2fb9478 in ResourceLoader::_load C:\Users\alula\workspace\godot\core\io\resource_loader.cpp:223
    #7 0x7ff7b2fba5a3 in ResourceLoader::_thread_load_function C:\Users\alula\workspace\godot\core\io\resource_loader.cpp:250
    #8 0x7ff7b2920fa7 in Thread::callback C:\Users\alula\workspace\godot\core\os\thread.cpp:75
    #9 0x7ff7b2924e18 in std::invoke<void (*)(Thread *, const Thread::Settings &, void (*)(void *), void *),Thread *,Thread::Settings,void (*)(void *),void *> C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31627\include\type_traits:1558
    #10 0x7ff7b292305d in std::thread::_Invoke<std::tuple<void (*)(Thread *, const Thread::Settings &, void (*)(void *), void *),Thread *,Thread::Settings,void (*)(void *),void *>,0,1,2,3,4> C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31627\include\thread:55
    #11 0x7ff7b6d7f909 in thread_start<unsigned int (__cdecl*)(void *),1> minkernel\crts\ucrt\src\appcrt\startup\thread.cpp:97
    #12 0x7ff7b6ce1719 in __asan::AsanThread::ThreadStart D:\a\_work\1\s\src\vctools\asan\llvm\compiler-rt\lib\asan\asan_thread.cpp:262
    #13 0x7ffa3c41244c in BaseThreadInitThunk+0x1c (C:\WINDOWS\System32\KERNEL32.DLL+0x18001244c)
    #14 0x7ffa3c70df77 in RtlUserThreadStart+0x27 (C:\WINDOWS\SYSTEM32\ntdll.dll+0x18005df77)

Thread T75 created by T73 here:
    #0 0x7ff7b6ce010f in __asan_wrap_CreateThread D:\a\_work\1\s\src\vctools\asan\llvm\compiler-rt\lib\asan\asan_win.cpp:149
    #1 0x7ff7b6d7fcce in _beginthreadex minkernel\crts\ucrt\src\appcrt\startup\thread.cpp:209
    #2 0x7ff7b2922a84 in std::thread::_Start<void (*)(Thread *, const Thread::Settings &, void (*)(void *), void *),Thread *,const Thread::Settings &,void (*&)(void *),void *&> C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31627\include\thread:75
    #3 0x7ff7b29225e8 in std::thread::thread<void (*)(Thread *, const Thread::Settings &, void (*)(void *), void *),Thread *,const Thread::Settings &,void (*&)(void *),void *&,0> C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31627\include\thread:90
    #4 0x7ff7b29213eb in Thread::start C:\Users\alula\workspace\godot\core\os\thread.cpp:91
    #5 0x7ff7b2fbc871 in ResourceLoader::load_threaded_request C:\Users\alula\workspace\godot\core\io\resource_loader.cpp:397
    #6 0x7ff7abb26d2a in ResourceLoaderText::load C:\Users\alula\workspace\godot\scene\resources\resource_format_text.cpp:468
    #7 0x7ff7abb59971 in ResourceFormatLoaderText::load C:\Users\alula\workspace\godot\scene\resources\resource_format_text.cpp:1697
    #8 0x7ff7b2fb9478 in ResourceLoader::_load C:\Users\alula\workspace\godot\core\io\resource_loader.cpp:223
    #9 0x7ff7b2fba5a3 in ResourceLoader::_thread_load_function C:\Users\alula\workspace\godot\core\io\resource_loader.cpp:250
    #10 0x7ff7b2920fa7 in Thread::callback C:\Users\alula\workspace\godot\core\os\thread.cpp:75
    #11 0x7ff7b2924e18 in std::invoke<void (*)(Thread *, const Thread::Settings &, void (*)(void *), void *),Thread *,Thread::Settings,void (*)(void *),void *> C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31627\include\type_traits:1558
    #12 0x7ff7b292305d in std::thread::_Invoke<std::tuple<void (*)(Thread *, const Thread::Settings &, void (*)(void *), void *),Thread *,Thread::Settings,void (*)(void *),void *>,0,1,2,3,4> C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31627\include\thread:55
    #13 0x7ff7b6d7f909 in thread_start<unsigned int (__cdecl*)(void *),1> minkernel\crts\ucrt\src\appcrt\startup\thread.cpp:97
    #14 0x7ff7b6ce1719 in __asan::AsanThread::ThreadStart D:\a\_work\1\s\src\vctools\asan\llvm\compiler-rt\lib\asan\asan_thread.cpp:262
    #15 0x7ffa3c41244c in BaseThreadInitThunk+0x1c (C:\WINDOWS\System32\KERNEL32.DLL+0x18001244c)
    #16 0x7ffa3c70df77 in RtlUserThreadStart+0x27 (C:\WINDOWS\SYSTEM32\ntdll.dll+0x18005df77)

Thread T73 created by T0 here:
    #0 0x7ff7b6ce010f in __asan_wrap_CreateThread D:\a\_work\1\s\src\vctools\asan\llvm\compiler-rt\lib\asan\asan_win.cpp:149
    #1 0x7ff7b6d7fcce in _beginthreadex minkernel\crts\ucrt\src\appcrt\startup\thread.cpp:209
    #2 0x7ff7b2922a84 in std::thread::_Start<void (*)(Thread *, const Thread::Settings &, void (*)(void *), void *),Thread *,const Thread::Settings &,void (*&)(void *),void *&> C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31627\include\thread:75
    #3 0x7ff7b29225e8 in std::thread::thread<void (*)(Thread *, const Thread::Settings &, void (*)(void *), void *),Thread *,const Thread::Settings &,void (*&)(void *),void *&,0> C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31627\include\thread:90
    #4 0x7ff7b29213eb in Thread::start C:\Users\alula\workspace\godot\core\os\thread.cpp:91
    #5 0x7ff7b2fbc871 in ResourceLoader::load_threaded_request C:\Users\alula\workspace\godot\core\io\resource_loader.cpp:397
    #6 0x7ff7b5323aa4 in core_bind::ResourceLoader::load_threaded_request C:\Users\alula\workspace\godot\core\core_bind.cpp:50
    #7 0x7ff7b53f544f in call_with_variant_args_ret_helper<core_bind::ResourceLoader,Error,const String &,const String &,bool,0,1,2> C:\Users\alula\workspace\godot\core\variant\binder_common.h:677
    #8 0x7ff7b53f50a1 in call_with_variant_args_ret_dv<core_bind::ResourceLoader,Error,const String &,const String &,bool> C:\Users\alula\workspace\godot\core\variant\binder_common.h:490
    #9 0x7ff7b53f4a12 in MethodBindTR<core_bind::ResourceLoader,Error,const String &,const String &,bool>::call C:\Users\alula\workspace\godot\core\object\method_bind.h:479
    #10 0x7ff7a14878b3 in GDScriptFunction::call C:\Users\alula\workspace\godot\modules\gdscript\gdscript_vm.cpp:1632
    #11 0x7ff7a1211a78 in GDScriptInstance::callp C:\Users\alula\workspace\godot\modules\gdscript\gdscript.cpp:1571
    #12 0x7ff7a802f181 in Node::_gdvirtual__process_call<0> C:\Users\alula\workspace\godot\scene\main\node.h:237
    #13 0x7ff7a7f720b3 in Node::_notification C:\Users\alula\workspace\godot\scene\main\node.cpp:56
    #14 0x7ff79ef3a4e9 in Node::_notificationv C:\Users\alula\workspace\godot\scene\main\node.h:45
    #15 0x7ff7b25c39b8 in Object::notification C:\Users\alula\workspace\godot\core\object\object.cpp:741
    #16 0x7ff7a7ad021d in SceneTree::_notify_group_pause C:\Users\alula\workspace\godot\scene\main\scene_tree.cpp:904
    #17 0x7ff7a7ad1fdb in SceneTree::process C:\Users\alula\workspace\godot\scene\main\scene_tree.cpp:454
    #18 0x7ff79e1d41f1 in Main::iteration C:\Users\alula\workspace\godot\main\main.cpp:2855
    #19 0x7ff79e0984c1 in OS_Windows::run C:\Users\alula\workspace\godot\platform\windows\os_windows.cpp:777
    #20 0x7ff79e061d1d in widechar_main C:\Users\alula\workspace\godot\platform\windows\godot_windows.cpp:177
    #21 0x7ff79e062196 in _main C:\Users\alula\workspace\godot\platform\windows\godot_windows.cpp:201
    #22 0x7ff79e06224c in main C:\Users\alula\workspace\godot\platform\windows\godot_windows.cpp:215
    #23 0x7ff79e0622e6 in WinMain C:\Users\alula\workspace\godot\platform\windows\godot_windows.cpp:228
    #24 0x7ff7b6d0f3c1 in __scrt_common_main_seh D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288
    #25 0x7ffa3c41244c in BaseThreadInitThunk+0x1c (C:\WINDOWS\System32\KERNEL32.DLL+0x18001244c)
    #26 0x7ffa3c70df77 in RtlUserThreadStart+0x27 (C:\WINDOWS\SYSTEM32\ntdll.dll+0x18005df77)

Thread T80 created by T75 here:
    #0 0x7ff7b6ce010f in __asan_wrap_CreateThread D:\a\_work\1\s\src\vctools\asan\llvm\compiler-rt\lib\asan\asan_win.cpp:149
    #1 0x7ff7b6d7fcce in _beginthreadex minkernel\crts\ucrt\src\appcrt\startup\thread.cpp:209
    #2 0x7ff7b2922a84 in std::thread::_Start<void (*)(Thread *, const Thread::Settings &, void (*)(void *), void *),Thread *,const Thread::Settings &,void (*&)(void *),void *&> C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31627\include\thread:75
    #3 0x7ff7b29225e8 in std::thread::thread<void (*)(Thread *, const Thread::Settings &, void (*)(void *), void *),Thread *,const Thread::Settings &,void (*&)(void *),void *&,0> C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31627\include\thread:90
    #4 0x7ff7b29213eb in Thread::start C:\Users\alula\workspace\godot\core\os\thread.cpp:91
    #5 0x7ff7b2fbc871 in ResourceLoader::load_threaded_request C:\Users\alula\workspace\godot\core\io\resource_loader.cpp:397
    #6 0x7ff7abb26d2a in ResourceLoaderText::load C:\Users\alula\workspace\godot\scene\resources\resource_format_text.cpp:468
    #7 0x7ff7abb59971 in ResourceFormatLoaderText::load C:\Users\alula\workspace\godot\scene\resources\resource_format_text.cpp:1697
    #8 0x7ff7b2fb9478 in ResourceLoader::_load C:\Users\alula\workspace\godot\core\io\resource_loader.cpp:223
    #9 0x7ff7b2fba5a3 in ResourceLoader::_thread_load_function C:\Users\alula\workspace\godot\core\io\resource_loader.cpp:250
    #10 0x7ff7b2920fa7 in Thread::callback C:\Users\alula\workspace\godot\core\os\thread.cpp:75
    #11 0x7ff7b2924e18 in std::invoke<void (*)(Thread *, const Thread::Settings &, void (*)(void *), void *),Thread *,Thread::Settings,void (*)(void *),void *> C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31627\include\type_traits:1558
    #12 0x7ff7b292305d in std::thread::_Invoke<std::tuple<void (*)(Thread *, const Thread::Settings &, void (*)(void *), void *),Thread *,Thread::Settings,void (*)(void *),void *>,0,1,2,3,4> C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31627\include\thread:55
    #13 0x7ff7b6d7f909 in thread_start<unsigned int (__cdecl*)(void *),1> minkernel\crts\ucrt\src\appcrt\startup\thread.cpp:97
    #14 0x7ff7b6ce1719 in __asan::AsanThread::ThreadStart D:\a\_work\1\s\src\vctools\asan\llvm\compiler-rt\lib\asan\asan_thread.cpp:262
    #15 0x7ffa3c41244c in BaseThreadInitThunk+0x1c (C:\WINDOWS\System32\KERNEL32.DLL+0x18001244c)
    #16 0x7ffa3c70df77 in RtlUserThreadStart+0x27 (C:\WINDOWS\SYSTEM32\ntdll.dll+0x18005df77)
```

### Steps to reproduce

I've been getting pretty random crashes while calling `load_threaded_get()` in my project - as turns out there's a data race in load_threaded_get which makes the pointer to semaphore no longer valid if the loading thread destroys it in meantime:

https://github.com/godotengine/godot/blob/72b5a4335e914997ef46928034e8371724b29280/core/io/resource_loader.cpp#L453-L482

**Comment by Calinou** [comment]
@alula Please upload a [minimal reproduction project](https://github.com/godotengine/godot/blob/master/CONTRIBUTING.md#provide-a-simple-example-project) to make this easier to troubleshoot.

Can you reproduce this in 4.0.alpha12?

**Comment by Anixias** [comment]
I am able to reproduce a similar issue in v4.0.rc2.mono.official [d2699dc7a] where the ResourceLoader gets stuck loading something forever seemingly randomly.

**Comment by YuriSizov** [comment]
@RandomShaper Do you think this is something that might've been fixed by you for 4.1 already?

**Comment by RandomShaper** [comment]
> @RandomShaper Do you think this is something that might've been fixed by you for 4.1 already?

Yes.

**Comment by YuriSizov** [comment]
Okay then, let's consider this fixed. If the issue persists, we can reopen it. If https://github.com/godotengine/godot/issues/78689 is related here, we can continue there as well.
